### ISSUE TITLE
Sway coredumping after 1.2 when running wayland native software

### ISSUE THREAD

**Issue by Mel34** [swaywm/sway#4527]
```
~ ▶ lsmod | grep nvidia | wc -l
0
```
I'm using RX580 with Arch linux Zen kernel 5.2.11

* Sway Version: 
sway version 1.2-rc1-18-gb556380d (Aug 19 2019, branch 'master')
_note: same happens with sway from arch official repositories (v1.2, wlroots 0.7)_

* Debug Log: 
https://gist.github.com/Mel34/8b7d1a377222dc80e9b4207cb37c54ea

* Configuration File:
https://gist.github.com/Mel34/a56ca05b66d6d631ce1436efbf7d7d4f
_note: same happens with default sway config_

* Steps to reproduce:
Run lutris
Run some gnome application (I tried with gedit, it crashed sway)
Run mpv with --gpu-context=wayland. 
Run firefox with MOZ_ENABLE_WAYLAND=1.
Run QT applications with QT_QPA_PLATFORM=wayland.
...
Basically almost all software that's wayland native crashes sway after 1.2 and wlroots 0.7. Exception is wldash.
If I export GDK_BACKEND=x11 and then run lutris it doesn't crash.
Anything that runs in xwayland works fine.

* Stack Trace, if sway crashes:
https://gist.github.com/Mel34/369721185b23727be193459090044e98

What's particularly troubling me is it seems nobody else is having these issues. I could really use some help diagnosing this.

**Comment by emersion** [comment]
Can't reproduce.

Can you try to compile with `-Db_sanitize=address` and upload logs?

**Comment by Mel34** [comment]
Sure, working on it.

**Comment by Mel34** [comment]
* Coredump:
https://gist.github.com/Mel34/cb15b30aef03760482ca5e379712b169
* Sway log:
https://gist.github.com/Mel34/eef03547de2b6910cce939cdc884337a

**Comment by emersion** [comment]
Doesn't seem like this was compiled with `-Db_sanitize=address`. This option will instrument the sway executable and print the detailed stack trace in the logs instead of producing a coredump.

**Comment by Mel34** [comment]
```
# Maintainer: Drew DeVault <sir@cmpwn.com>
pkgname=sway-git
_pkgname=sway
pkgver=r6038.53e01bf5
pkgrel=1
license=("MIT")
pkgdesc="i3-compatible Wayland compositor"
makedepends=("meson" "git" "scdoc" "wayland-protocols" "ninja")
depends=(
	"json-c" "pcre" "wlroots-git" "cairo" "pango" "gdk-pixbuf2" "pam"
	"xorg-server-xwayland" "swaybg-git"
)
optdepends=(
	"rxvt-unicode: Default terminal emulator."
	"dmenu: Default for launching applications."
	"i3status: To display system information with a bar."
	"swayidle-git: Idle management daemon for Sway."
	"swaylock-git: Screen locker for Sway."
)
arch=("i686" "x86_64")
url="https://swaywm.org"
source=("${pkgname%-*}::git+https://github.com/swaywm/sway.git")
sha512sums=('SKIP')
provides=("sway")
conflicts=("sway")
options=(debug !strip)

pkgver() {
	cd "$_pkgname"
	printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
}

prepare() {
	cd "$_pkgname"
}

build() {
	cd "$_pkgname"
	meson \
		-Dwerror=false \
		-Db_sanitize=address \
		--prefix /usr \
		"$srcdir/build"
	ninja -C "$srcdir/build"
}

package() {
	cd "$_pkgname"
	DESTDIR="$pkgdir" ninja -C "$srcdir/build" install
}
```
This is the PKGBUILD I used, can you please instruct me how to properly add `-Db_sanitize=address`?

**Comment by emersion** [comment]
```sh
git clone https://github.com/swaywm/sway
cd sway
meson build/ -Db_sanitize=address
ninja -C build
build/sway/sway
```

**Comment by Mel34** [comment]
Okay, while I do that, here's the lutris log from `WAYLAND_DEBUG=1 lutris`
https://gist.github.com/Mel34/cfddd9a9affb919cf9ca350ef8c721e2

**Comment by emersion** [comment]
magentar provided a ASan trace:

```
ibEGL warning: FIXME: egl/x11 doesn't support front buffer rendering.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==2091==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000079 (pc 0x7f0206042ac7 bp 0x7ffef98f18c0 sp 0x7ffef98f18a8 T0)
==2091==The signal is caused by a WRITE memory access.
==2091==Hint: address points to the zero page.
    #0 0x7f0206042ac6 in wl_list_insert (/lib64/libwayland-server.so.0+0xdac6)
    #1 0x449563 in wl_signal_add /usr/include/wayland-server-core.h:435
    #2 0x44c0ff in handle_map ../sway/desktop/xdg_shell.c:447
    #3 0x7f0205ff509a in wlr_signal_emit_safe ../util/signal.c:29
    #4 0x7f0205fce84b in handle_xdg_surface_commit ../types/xdg_shell/wlr_xdg_surface.c:360
    #5 0x7f0205fee063 in surface_commit_pending ../types/wlr_surface.c:372
    #6 0x7f0205fee2e1 in surface_commit ../types/wlr_surface.c:444
    #7 0x7f0205795b27 in ffi_call_unix64 (/lib64/libffi.so.6+0x6b27)
    #8 0x7f0205795338 in ffi_call (/lib64/libffi.so.6+0x6338)
    #9 0x7f02060421b6  (/lib64/libwayland-server.so.0+0xd1b6)
    #10 0x7f020603ed31  (/lib64/libwayland-server.so.0+0x9d31)
    #11 0x7f0206040369 in wl_event_loop_dispatch (/lib64/libwayland-server.so.0+0xb369)
    #12 0x7f020603ef24 in wl_display_run (/lib64/libwayland-server.so.0+0x9f24)
    #13 0x42c127 in server_run ../sway/server.c:202
    #14 0x42a9db in main ../sway/main.c:400
    #15 0x7f0205d48f32 in __libc_start_main (/lib64/libc.so.6+0x23f32)
    #16 0x40c0bd in _start (/home/frederick/Documents/Programs/sway/build/sway/sway+0x40c0bd)
 
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/lib64/libwayland-server.so.0+0xdac6) in wl_list_insert
==2091==ABORTING
2019-09-01 19:13:57 - [common/ipc-client.c:88] Unable to receive IPC response
```

**Comment by emersion** [comment]
I'm not sure what's wrong here.

Have you made sure to always use either `sway-git` and `wlroots-git` built at the same time, or the latest release of both? (ie. don't try to run `sway-git` with an old `wlroots-git`)

**Comment by Mel34** [comment]
I can't even execute your last step, `build/sway/sway`, sway sefgaults when doing that too...
https://gist.github.com/Mel34/f00a22b1b893ed537777dbfce0449f71

**Comment by Mel34** [comment]
Yes, here's what's installed:
```
~ ▶ pacman -Qi sway-git 
Name            : sway-git
Version         : r6038.53e01bf5-1
Description     : i3-compatible Wayland compositor
Architecture    : x86_64
URL             : https://swaywm.org
Licenses        : MIT
Groups          : None
Provides        : sway
Depends On      : json-c  pcre  wlroots-git  cairo  pango  gdk-pixbuf2  pam  xorg-server-xwayland  swaybg-git
Optional Deps   : rxvt-unicode: Default terminal emulator.
                  dmenu: Default for launching applications.
                  i3status: To display system information with a bar.
                  swayidle-git: Idle management daemon for Sway.
                  swaylock-git: Screen locker for Sway.
Required By     : None
Optional For    : None
Conflicts With  : sway
Replaces        : None
Installed Size  : 12,76 MiB
Packager        : Anon <anon@loonix>
Build Date      : nedjelja, 01. rujna 2019. 19:10:38 CEST
Install Date    : nedjelja, 01. rujna 2019. 19:10:59 CEST
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : None

~ ▶ pacman -Qi wlroots-git 
Name            : wlroots-git
Version         : 0.7.0.r1.g734c64a6-1
Description     : Modular Wayland compositor library
Architecture    : x86_64
URL             : https://github.com/swaywm/wlroots
Licenses        : custom:MIT
Groups          : None
Provides        : wlroots
Depends On      : libcap  systemd  wayland  opengl-driver  libxcb  xcb-util-errors  xcb-util-image  xcb-util-wm  pixman  libinput  libxkbcommon  libpng  ffmpeg
Optional Deps   : None
Required By     : sway-git
Optional For    : None
Conflicts With  : wlroots
Replaces        : None
Installed Size  : 3,72 MiB
Packager        : Anon <anon@loonix>
Build Date      : nedjelja, 01. rujna 2019. 00:05:43 CEST
Install Date    : nedjelja, 01. rujna 2019. 00:29:11 CEST
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : SHA-256 Sum
```

**Comment by emersion** [comment]
Looks good.

This will require manual debugging. Start `gdb sway` in another X11/Wayland session, then try to understand what's going on (e.g. `frame 2`, `info locals`, `p xdg_surface->toplevel`, etc).

**Comment by Mel34** [comment]
So I started `gdb sway` and issues run command, this is what I get:
```
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /usr/local/bin/sway 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[Detaching after vfork from child process 26160]
[New Thread 0x7fffee602700 (LWP 26211)]
[New Thread 0x7fffedcac700 (LWP 26213)]
[New Thread 0x7fffed4ab700 (LWP 26214)]
[New Thread 0x7fffeccaa700 (LWP 26215)]
[New Thread 0x7fffdffff700 (LWP 26216)]
[New Thread 0x7fffdf7fe700 (LWP 26217)]
[New Thread 0x7fffdeffd700 (LWP 26218)]
[New Thread 0x7fffde7fc700 (LWP 26220)]

Thread 2 "sway:cs0" received signal SIG33, Real-time event 33.
[Switching to Thread 0x7fffee602700 (LWP 26211)]
0x00007ffff76c3415 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
```

**Comment by emersion** [comment]
`continue` should do it

**Comment by Mel34** [comment]
Here's a log of that session, I tried running lutris from inside that sway session:
https://gist.github.com/Mel34/0075f77c726c7668ac7e4949c8da7f3f
At some point I lost my cursor in xfce session, it came back after I used `continue` in gdb.

**Comment by emersion** [comment]
The idea is to make sway crash, then to inspect the process' state (no more `continue` commands).

**Comment by Mel34** [comment]
I've never used gdb before so thank you for your patience with me. I can restart the process and get it to crash, what then?

**Comment by emersion** [comment]
After the crash, you'll need to inspect sway's state and understand why it crashes. An example of things I'd try might be: `frame 2` (selects the second frame in the stack trace, please have a look at `sway/desktop/xdg_shell.c:447`), `info locals` (prints all local variables), `p xdg_surface->toplevel` (prints a variable that might be NULL).

The goal is to understand where this NULL is coming from.

**Comment by Mel34** [comment]
```
(gdb) frame 2
#2  0x000055555557dd83 in handle_map (listener=0x555555b6ba10, data=0x555555b6b450) at ../sway/desktop/xdg_shell.c:447
447             wl_signal_add(&xdg_surface->toplevel->events.request_fullscreen,
(gdb) info locals
xdg_shell_view = 0x555555b6b860
view = 0x555555b6b860
xdg_surface = 0x555555b6b450
csd = true
(gdb) p xdg_surface->toplevel
$1 = (struct wlr_xdg_toplevel *) 0x555555b6b630
```
Input requires multiple keypresses, and I have no mouse. This is tedious. I issues a `quit` command just to be able to reply to you here.

**Comment by Mel34** [comment]
I managed to run sway from local build, lutris started up normally, MOZ_ENABLE_WAYLAND=1 firefox also. I have no idea what's going on here anymore. I guess I'll compare the build process to what the PKGBUILD does and try to find the differences.

**Comment by Mel34** [comment]
You might be interested in knowing I traced it to 
```
meson \
		-Dwerror=false \
```
that was in PKGBUILD. Removing `-Dwerror=false\` fixes the issue for me.
